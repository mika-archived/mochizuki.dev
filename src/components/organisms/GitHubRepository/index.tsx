import React, { useEffect, useRef, useState } from "react";
import styled from "styled-components";

import { Small } from "../../atoms/Typography";
import LabeledIcon from "../../molecules/LabeledIcon";
import LanguageBadge from "../../molecules/LanguageBadge";
import LinkLabeledIcon from "../../molecules/LinkLabeledIcon";

type Props = {
  nameWithOwner: string;
  description: string;
  forks: { totalCount: number };
  primaryLanguage: { color: string; name: string };
  stargazers: { totalCount: number };
};

const Border = styled.div`
  display: flex;
  flex-flow: column nowrap;
  height: calc(100% - 32px);
  max-height: 100%;
  padding: 16px;
  border: 1px solid #ccc;
  border-radius: 4px;
`;

const NameWithOwner = styled(LinkLabeledIcon)`
  color: #333;
`;

const Description = styled.div<{ lines: number }>`
  /* stylelint-disable-next-line value-no-vendor-prefix */
  display: -webkit-box;
  flex: 1 1 auto;
  margin: 5px 0;
  overflow: hidden;
  line-height: 1.3;
  color: #888;
  text-overflow: ellipsis;

  /* 
   * multiline overflow: ellipsis does not supported in currently
   * but, -webkit-line-clamp has support by all of modern browsers.
   */
  /* stylelint-disable-next-line property-no-vendor-prefix */
  -webkit-box-orient: vertical;
  /* stylelint-disable-next-line property-no-vendor-prefix */
  -webkit-line-clamp: ${props => props.lines};
`;

const Forks = styled(LabeledIcon)`
  margin: 1px 0 0 10px;
  color: #888;
`;

const Metadata = styled.div`
  display: flex;
  flex-flow: row nowrap;
  font-size: 14px;
`;

const Stargazers = styled(LabeledIcon)`
  margin: 1px 0 0 10px;
  color: #888;
`;

const GitHubRepository: React.FC<Props> = ({ description, forks, nameWithOwner, primaryLanguage, stargazers }) => {
  const box = useRef<HTMLDivElement>(null);
  const [lines, setLines] = useState<number>(-1);

  useEffect(() => {
    const boxHeight = box.current?.clientHeight!;
    const lineHeight = 14 * 1.2;

    setLines(Math.floor(boxHeight / lineHeight));
  });

  return (
    <Border>
      <NameWithOwner href={`https://github.com/${nameWithOwner}`} icon="book" prefix="light" label={nameWithOwner} />
      <Description ref={box} lines={lines}>
        <Small>{description}</Small>
      </Description>
      <Metadata>
        <LanguageBadge color={primaryLanguage.color} language={primaryLanguage.name} />
        <Stargazers icon="star" prefix="solid" label={`${stargazers.totalCount}`} />
        <Forks icon="code-branch" prefix="light" label={`${forks.totalCount}`} />
      </Metadata>
    </Border>
  );
};

export default GitHubRepository;
