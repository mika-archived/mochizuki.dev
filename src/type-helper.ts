import { StyledComponent } from "styled-components";

// Extract Props of React Components
export type ExtractProps<T> =
  // prettier-ignore
  T extends StyledComponent<any, any, infer R, never> ? R
  : T extends React.FunctionComponent<infer R> ? R
  : string;

export type ReactClassName = {
  className?: string;
};
