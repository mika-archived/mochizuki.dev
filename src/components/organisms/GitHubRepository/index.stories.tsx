import React from "react";
import styled from "styled-components";
import { storiesOf } from "@storybook/react";

import { ExtractProps } from "../../../type-helper";

import GitHubRepository from ".";

const repository: ExtractProps<typeof GitHubRepository> = {
  nameWithOwner: "mika-sandbox/dotnet-window-capture",
  description: "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
  forks: { totalCount: 1 },
  primaryLanguage: { color: "#178600", name: "C#" },
  stargazers: { totalCount: 12 }
};

const LimitedBox1 = styled.div`
  width: 290px;
  height: 140px;
  margin: 10px 0;
`;

const LimitedBox2 = styled.div`
  width: 290px;
  height: 120px;
  margin: 10px 0;
`;

storiesOf("organisms/GitHubRepository", module).add("default", () => (
  <>
    {/* eslint-disable react/jsx-props-no-spreading */}
    <LimitedBox1>
      <GitHubRepository {...repository} />
    </LimitedBox1>
    <LimitedBox1>
      <GitHubRepository {...repository} />
    </LimitedBox1>
    <LimitedBox2>
      <GitHubRepository {...repository} />
    </LimitedBox2>
  </>
));
