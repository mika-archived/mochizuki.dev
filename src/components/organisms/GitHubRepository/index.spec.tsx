import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

import React from "react";
import { cleanup, render } from "@testing-library/react";

import { ExtractProps } from "../../../type-helper";

import GitHubRepository from ".";

const repository: ExtractProps<typeof GitHubRepository> = {
  nameWithOwner: "mika-sandbox/dotnet-window-capture",
  description: "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
  forks: { totalCount: 1 },
  primaryLanguage: { color: "#178600", name: "C#" },
  stargazers: { totalCount: 12 }
};

afterEach(cleanup);

test("does not have changes from snapshot", () => {
  // eslint-disable-next-line react/jsx-props-no-spreading
  const { asFragment } = render(<GitHubRepository {...repository} />);

  expect(asFragment()).toMatchSnapshot();
});
