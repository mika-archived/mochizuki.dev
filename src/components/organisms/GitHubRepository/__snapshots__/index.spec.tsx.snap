// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not have changes from snapshot 1`] = `
<DocumentFragment>
  .c6 {
  margin: 0;
  font-family: "游ゴシック体","Yu Gothic",YuGothic,"ヒラギノ角ゴシック Pro","Hiragino Kaku Gothic Pro",メイリオ,Meiryo,sans-serif;
  font-size: 14px;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
}

.c14 {
  margin: 0;
  font-family: "游ゴシック体","Yu Gothic",YuGothic,"ヒラギノ角ゴシック Pro","Hiragino Kaku Gothic Pro",メイリオ,Meiryo,sans-serif;
  margin-left: 4px;
  overflow: hidden;
  line-height: 1.2;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c9 {
  background-color: #178600;
  border-radius: 9999px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
}

.c10 {
  width: 16px;
  height: 16px;
}

.c11 {
  margin: 0;
  font-family: "游ゴシック体","Yu Gothic",YuGothic,"ヒラギノ角ゴシック Pro","Hiragino Kaku Gothic Pro",メイリオ,Meiryo,sans-serif;
  font-size: 14px;
  margin-left: 4px;
  line-height: 1.3;
  color: #888;
}

.c4 {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
}

.c3 {
  margin: 0;
  font-family: "游ゴシック体","Yu Gothic",YuGothic,"ヒラギノ角ゴシック Pro","Hiragino Kaku Gothic Pro",メイリオ,Meiryo,sans-serif;
  margin-left: 4px;
  overflow: hidden;
  line-height: 1.2;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  height: calc(100% - 32px);
  max-height: 100%;
  padding: 16px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.c2 {
  color: #333;
}

.c5 {
  display: -webkit-box;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  margin: 5px 0;
  overflow: hidden;
  line-height: 1.3;
  color: #888;
  text-overflow: ellipsis;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 0;
}

.c15 {
  margin-left: 10px;
  color: #888;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  font-size: 14px;
}

.c13 {
  margin-left: 10px;
  color: #888;
}

<div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <i
        class="fal fa-book fa-fw"
      />
      <p
        class="sc-AxjAm c3"
      >
        <a
          class="c4"
          href="https://github.com/mika-sandbox/dotnet-window-capture"
          rel="noreferrer noopener"
          target="_blank"
        >
          mika-sandbox/dotnet-window-capture
        </a>
      </p>
    </div>
    <div
      class="c5"
    >
      <p
        class="sc-AxjAm c6"
      >
        Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </p>
    </div>
    <div
      class="c7"
    >
      <div
        class="c8"
      >
        <div
          class="c9 c10"
          color="#178600"
        />
        <p
          class="sc-AxjAm c11"
        >
          C#
        </p>
      </div>
      <div
        class="c12 c13"
      >
        <i
          class="fas fa-star fa-fw"
        />
        <p
          class="sc-AxjAm c14"
        >
          12
        </p>
      </div>
      <div
        class="c12 c15"
      >
        <i
          class="fal fa-code-branch fa-fw"
        />
        <p
          class="sc-AxjAm c14"
        >
          1
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
